src/index/VocabularyGenerator.cpp

```diff
@@ -75,86 +70,166 @@ size_t mergeVocabulary(const std::string& basename, size_t numFiles,
     }
   }
 
-  // keep track of the last seen word to correctly handle duplicates
-  std::string lastWritten = "";
-  // the number of words we have written. This also is the global Id of the next
-  // word we see, unless it is is equal to the previous word
-  size_t totalWritten = 0;
-  bool firstLangPredSeen = false;
-  *langPredLowerBound = 0;
-  *langPredUpperBound = 0;
+  std::vector<QueueWord> sortedBuffer;
+  sortedBuffer.reserve(_bufferSize + 2);
+
+  std::future<void> writeFuture;
 
   // start k-way merge
   while (!queue.empty()) {
-    auto top = queue.top();
+    sortedBuffer.push_back(std::move(queue.top()));
     queue.pop();
+    auto i = sortedBuffer.back()._partialFileId;
+
+    if (sortedBuffer.size() >= _bufferSize) {
+      auto task = [this, buf = std::move(sortedBuffer)]() {
+        this->writeQueueWordsToIdVec(buf);
+      };
+      sortedBuffer.clear();
+      sortedBuffer.reserve(_bufferSize + 2);
+      // wait for the last batch
+
+      if (writeFuture.valid()) {
+        LOG(INFO) << "Waiting for the asynchronous write to finish\n";
+        writeFuture.get();
+      }
+      writeFuture = std::async(task);
+      // we have moved away our buffer, start over
+    }
+
+    // add next word from the same infile to the priority queue
+    if (endOfFile[i]) {
+      continue;
+    }  // file is exhausted, nothing to add
+
+    endOfFile[i] = true;
+    uint32_t len;
+    if (infiles[i].read((char*)&len, sizeof(len))) {
+      std::string word(len, '\0');
+      infiles[i].read(&(word[0]), len);
+      Id id;
+      infiles[i].read((char*)&id, sizeof(id));
+      queue.push(QueueWord(word, i, id));
```

Comment:
This moves the `word` into `QueueWord` right? Maybe that should have an explicit `std::move()`

Reply:
that was old code, but good find! I think with the current parallelism this is not a bottleneck, but we definitely want to move here.